### PR TITLE
Create GuildMemberUpdateEvent hierarchy

### PR DESCRIPTION
**Pull request by AtriusX**
[contributing]: https://github.com/DV8FromTheWorld/JDA/wiki/5%29-Contributing

## Pull Request Etiquette

- [x] I have checked the PRs for upcoming features/bug fixes.
- [x] I have read the [contributing guidelines][contributing].

### Changes

- [ ] Internal code
- [x] Library interface (affecting end-user code) 
- [ ] Documentation
- [ ] Other: \_____ 

Closes Issue: #973 

## Description

This adds a new update event for guild members, and refactors the nickname update event to extend it.

### CODE REVIEW

File: src/main/java/net/dv8tion/jda/api/events/guild/member/update/GuildMemberUpdateNicknameEvent.java [hunk 1]
```diff
@@ -30,17 +31,17 @@
  *
  * <p>Identifier: {@code nick}
  */
-public class GuildMemberUpdateNicknameEvent extends GenericGuildMemberEvent implements UpdateEvent<Member, String>
+public class GuildMemberUpdateNicknameEvent extends GenericGuildMemberUpdateEvent<String>
 {
     public static final String IDENTIFIER = "nick";
 
     private final String oldNick, newNick;
```
Comment (DV8FromTheWorld) [inline src/main/java/net/dv8tion/jda/api/events/guild/member/update/GuildMemberUpdateNicknameEvent.java]: These can be removed now

File: src/main/java/net/dv8tion/jda/api/events/guild/member/update/GuildMemberUpdateNicknameEvent.java [hunk 2]
```diff
@@ -30,17 +31,17 @@
...
 
     public GuildMemberUpdateNicknameEvent(@Nonnull JDA api, long responseNumber, @Nonnull Member member, @Nullable String oldNick)
     {
-        super(api, responseNumber, member);
+        super(api, responseNumber, member, oldNick, member.getNickname(), IDENTIFIER);
         this.oldNick = oldNick;
-        this.newNick = member.getNickname();
+        this.newNick = super.getNewValue();
     }
```
Comment (DV8FromTheWorld) [inline src/main/java/net/dv8tion/jda/api/events/guild/member/update/GuildMemberUpdateNicknameEvent.java]: All methods besides `getOldNickname` and `getNewNickname` can be removed.

File: src/main/java/net/dv8tion/jda/api/hooks/ListenerAdapter.java [hunk 3]
```diff
@@ -279,6 +283,7 @@ public void onGenericRole(@Nonnull GenericRoleEvent event) {}
     public void onGenericRoleUpdate(@Nonnull GenericRoleUpdateEvent event) {}
     public void onGenericEmote(@Nonnull GenericEmoteEvent event) {}
     public void onGenericEmoteUpdate(@Nonnull GenericEmoteUpdateEvent event) {}
+    public void onGenericGuildMemberUpdateEvent(@Nonnull GenericGuildMemberUpdateEvent event) {}
```
Comment (DV8FromTheWorld) [inline src/main/java/net/dv8tion/jda/api/hooks/ListenerAdapter.java]: Move this to be just below `onGenericGuildMember`

Comment (DV8FromTheWorld) [reply src/main/java/net/dv8tion/jda/api/hooks/ListenerAdapter.java]: Also, remove `Event` from the method name

File: src/main/java/net/dv8tion/jda/api/hooks/ListenerAdapter.java [hunk 4]
```diff
@@ -573,6 +580,8 @@ else if (event instanceof GenericVoiceChannelUpdateEvent)
             onGenericVoiceChannelUpdate((GenericVoiceChannelUpdateEvent) event);
         else if (event instanceof GenericGuildUpdateEvent)
             onGenericGuildUpdate((GenericGuildUpdateEvent) event);
+        else if (event instanceof GenericGuildMemberUpdateEvent)
+            onGenericGuildMemberUpdateEvent(((GenericGuildMemberUpdateEvent) event));
```
Comment (DV8FromTheWorld) [inline src/main/java/net/dv8tion/jda/api/hooks/ListenerAdapter.java]: Remove the extra, unneeded enclosing parentheses.

File: src/main/java/net/dv8tion/jda/api/hooks/ListenerAdapter.java [hunk 5]
```diff
@@ -573,6 +580,8 @@ else if (event instanceof GenericVoiceChannelUpdateEvent)
...
         else if (event instanceof GenericGuildMemberEvent)
             onGenericGuildMember((GenericGuildMemberEvent) event);
```
Comment (DV8FromTheWorld) [inline src/main/java/net/dv8tion/jda/api/hooks/ListenerAdapter.java]: The check for `onGenericGuildMember` needs to be moved to the if-block for generic superclasses.

It starts under `//Generic events that have generic subclasses (the subclasses as above).`

File: src/main/java/net/dv8tion/jda/api/hooks/ListenerAdapter.java (hunk 3 above)
Comment (AtriusX) [reply src/main/java/net/dv8tion/jda/api/hooks/ListenerAdapter.java]: Done

File: src/main/java/net/dv8tion/jda/api/events/guild/member/update/GuildMemberUpdateNicknameEvent.java (hunk 2 above)
Comment (AtriusX) [reply src/main/java/net/dv8tion/jda/api/events/guild/member/update/GuildMemberUpdateNicknameEvent.java]: Alright

File: src/main/java/net/dv8tion/jda/api/hooks/ListenerAdapter.java (hunk 5 above)
Comment (AtriusX) [reply src/main/java/net/dv8tion/jda/api/hooks/ListenerAdapter.java]: Thanks for the heads up! Moved it.

File: src/main/java/net/dv8tion/jda/api/hooks/ListenerAdapter.java (hunk 4 above)
Comment (AtriusX) [reply src/main/java/net/dv8tion/jda/api/hooks/ListenerAdapter.java]: Didn't notice those, thanks

File: src/main/java/net/dv8tion/jda/api/events/guild/member/update/GenericGuildMemberUpdateEvent.java [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+
+/**
+ * Indicates that a {@link net.dv8tion.jda.api.entities.Guild Guild} member event is fired.
+ * <br>Every GuildMemberUpdateEvent is an instance of this event and can be casted.
+ *
+ * <p>Can be used to detect any GuildMemberUpdateEvent.
+ */
+public abstract class GenericGuildMemberUpdateEvent<T> extends GenericGuildMemberEvent implements UpdateEvent<Member, T> {
```
Comment (MinnDevelopment) [inline src/main/java/net/dv8tion/jda/api/events/guild/member/update/GenericGuildMemberUpdateEvent.java]: Braces on new line

File: src/main/java/net/dv8tion/jda/api/events/guild/member/update/GenericGuildMemberUpdateEvent.java [hunk 7]
```diff
@@ -0,0 +1,81 @@
...
+
+    protected final T previous;
+    protected final T next;
+    protected final String identifier;
+
+    public GenericGuildMemberUpdateEvent(
+        @Nonnull JDA api, long responseNumber, @Nonnull Member member,
+        @Nullable T previous, @Nullable T next, String identifier)
```
Comment (MinnDevelopment) [inline src/main/java/net/dv8tion/jda/api/events/guild/member/update/GenericGuildMemberUpdateEvent.java]: The identifier is also nonnull

**Comment by AtriusX** [comment]
Sorry about that! Should be fixed now.

**Pull request merged**